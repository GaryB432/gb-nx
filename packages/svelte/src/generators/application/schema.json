{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "GbNxSvelteApplication",
  "title": "Add NX configuration to your Svelte-kit package.",
  "description": "Add NX configuration to your Svelte-kit package.",
  "examples": [
    {
      "command": "nx g @gb-nx/svelte:application myapp",
      "description": "Adds nx key to the svelte kit package.json in the workspace's myapp project"
    }
  ],
  "type": "object",
  "properties": {
    "projectPath": {
      "type": "string",
      "description": "Path to the Sveltekit project.",
      "examples": ["apps/my-web"],
      "x-priority": "important"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "eslint": {
      "type": "boolean",
      "description": "Update workspace eslint to accomodate Svelte linting",
      "default": true
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    }
  },
  "additionalProperties": false,
  "required": ["projectPath"]
}
