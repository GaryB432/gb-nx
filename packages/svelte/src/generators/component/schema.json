{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "GbNxSvelteComponent",
  "title": "Add a component to a Svelte Kit project",
  "description": "Svelte Component Options Schema.",
  "examples": [
    {
      "command": "nx g @gb-nx/svelte:component admin --project myapp",
      "description": "Add a Admin component to the myapp Svelte-Kit application"
    },
    {
      "command": "nx g @gb-nx/svelte:component admin --project myapp --language ts",
      "description": "Add a Typescript language Admin component to the myapp Svelte-Kit application"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the component.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "project": {
      "description": "The Svelte project to target.",
      "type": "string",
      "$default": {
        "$source": "projectName"
      },
      "alias": "p",
      "x-prompt": "What Svelte project would you like to target?"
    },
    "directory": {
      "description": "Directory where the component is placed, relative to your Svelte project lib folder.",
      "type": "string",
      "aliases": ["dir", "path"]
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false
    },
    "language": {
      "description": "Component script language (ts/js).",
      "type": "string",
      "enum": ["js", "ts"],
      "default": "js"
    },
    "style": {
      "description": "Component style language (css/scss).",
      "type": "string",
      "enum": ["css", "scss"],
      "default": "css"
    }
  },
  "additionalProperties": false,
  "required": ["name", "project"]
}
