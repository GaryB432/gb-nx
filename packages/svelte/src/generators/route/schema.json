{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "GbNxSvelteRoute",
  "title": "Create a route in a Svelte Kit project.",
  "description": "Create a route in a Svelte Kit project.",
  "examples": [
    {
      "command": "nx g @gb-nx/svelte:route myroute --project myapp --language scss",
      "description": "Generate a route in the myapp Svelte Kit package"
    },
    {
      "command": "nx g @gb-nx/svelte:route myroute --project myapp --load universal",
      "description": "Generate a route with a load function to run on both client and server"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the route.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "project": {
      "description": "The Svelte project to target.",
      "type": "string",
      "$default": {
        "$source": "projectName"
      },
      "alias": "p"
    },
    "directory": {
      "description": "Directory where the generated files are placed, relative to your Svelte project's source root.",
      "type": "string",
      "aliases": ["dir", "path"]
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false
    },
    "skipTests": {
      "description": "Do not create 'spec' test file for the route.",
      "type": "boolean",
      "default": false
    },
    "language": {
      "description": "Route script language (ts/js).",
      "type": "string",
      "enum": ["js", "ts"],
      "default": "js"
    },
    "style": {
      "description": "Route style language (css/scss).",
      "type": "string",
      "enum": ["css", "scss"],
      "default": "css"
    },
    "load": {
      "description": "Should the load function run on client and server (universal) or just server",
      "type": "string",
      "enum": ["universal", "server", "none"],
      "default": "none"
    },
    "runes": {
      "description": "Use svelte runes (requires svelte >=5)",
      "type": "boolean"
    }
  },
  "additionalProperties": false,
  "required": ["name"]
}
