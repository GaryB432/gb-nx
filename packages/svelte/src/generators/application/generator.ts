import {
  NX_VERSION,
  addDependenciesToPackageJson,
  ensurePackage,
  formatFiles,
  installPackagesTask,
  joinPathFragments,
  normalizePath,
  updateJson,
  writeJson,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';

import { initGenerator } from '@nx/js';

import {
  eslintVersion,
  typescriptESLintVersion,
} from '@nx/eslint/src/utils/versions';
import type {
  NxProjectPackageJsonConfiguration,
  PackageJson,
} from 'nx/src/utils/package-json';
import { updateEslint } from '../../utils/eslint';
import { includes } from '../../utils/globber';
import { isSvelte } from '../../utils/svelte';
import {
  eslintPluginGbVersion,
  eslintPluginSvelteVersion,
  prettierPluginSvelteVersion,
} from '../../utils/versions';
import { normalizeOptions } from './lib/normalize-options';
import { type Config as PrettierConfig } from './lib/prettier';
import {
  type ApplicationGeneratorOptions,
  type NormalizedOptions,
} from './schema';

const PRETTIER_PLUGIN_SVELTE = 'prettier-plugin-svelte';

const nx: NxProjectPackageJsonConfiguration = {
  namedInputs: {
    default: ['{projectRoot}/**/*'],
    production: [
      '!{projectRoot}/.svelte-kit/*',
      '!{projectRoot}/build/*',
      '!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)',
      '!{projectRoot}/tsconfig.spec.json',
    ],
  },
  targets: {
    build: {
      inputs: ['production', '^production'],
      outputs: ['{projectRoot}/build'],
      dependsOn: ['^build'],
    },
  },
};

function addNxConfig(tree: Tree, packageJsonPath: string): void {
  return updateJson<PackageJson>(tree, packageJsonPath, (json) => ({
    ...json,
    nx,
  }));
}

function addScriptsToPackageJson(
  tree: Tree,
  scripts: Record<string, string>,
  packageJsonPath: string
): void {
  updateJson<PackageJson>(tree, packageJsonPath, (json) => {
    json.scripts = json.scripts ?? {};
    for (const script of Object.keys(scripts)) {
      json.scripts[script] = scripts[script];
    }
    return json;
  });
}

function addWorkspaceToPackageJson(
  tree: Tree,
  options: NormalizedOptions,
  packageJsonPath = 'package.json'
): void {
  updateJson<PackageJson>(tree, packageJsonPath, (json) => {
    json.workspaces = json.workspaces ?? [];
    if (Array.isArray(json.workspaces)) {
      if (!includes(options.projectRoot, json.workspaces)) {
        json.workspaces.push(normalizePath(options.projectRoot));
        json.workspaces.sort();
      }
    } else {
      throw new Error('TODO handle');
    }
    return json;
  });
}

function updatePrettier(tree: Tree, options: NormalizedOptions) {
  ensurePackage(PRETTIER_PLUGIN_SVELTE, prettierPluginSvelteVersion);

  const updateConfig = () => {
    const prettierrc = '.prettierrc';
    if (!tree.exists(prettierrc)) {
      writeJson(tree, prettierrc, { singleQuote: true });
    }

    updateJson<PrettierConfig, PrettierConfig>(tree, prettierrc, (json) => {
      json.plugins = json.plugins ?? [];
      if (!json.plugins.includes(PRETTIER_PLUGIN_SVELTE)) {
        json.plugins.push(PRETTIER_PLUGIN_SVELTE);
      }
      json.overrides = json.overrides ?? [];
      if (!json.overrides.some((t) => t.files === '*.svelte')) {
        json.overrides.push({
          files: '*.svelte',
          options: { parser: 'svelte' },
        });
      }
      return json;
    });
  };
  const updateIgnore = () => {
    const fname = '.prettierignore';
    const tbs = ['.svelte-kit', 'build'];
    const newPatterns = tbs.map((p) =>
      joinPathFragments(options.projectRoot, p)
    );

    let content = [
      '# Add files here to ignore them from prettier formatting',
      '/dist',
      '/coverage',
      '/.nx/cache',
    ].join('\n');

    if (tree.exists(fname)) {
      content = tree.read(fname, 'utf-8')!;
    }

    const patterns = content.split('\n');

    const lines = patterns
      .concat(newPatterns.filter((p) => !patterns.includes(p)))
      .concat('');

    tree.write(fname, lines.join('\n'));
  };
  addDependenciesToPackageJson(
    tree,
    {},
    {
      'prettier-plugin-svelte': prettierPluginSvelteVersion,
    }
  );
  updateConfig();
  updateIgnore();
}

export default async function (
  tree: Tree,
  options: ApplicationGeneratorOptions
): Promise<GeneratorCallback> {
  const notSvelte = (p: string) =>
    `project at '${p}' is not configured for svelte`;

  const normalizedOptions = await normalizeOptions(tree, options);
  const config = { root: normalizedOptions.projectRoot };

  if (!isSvelte(tree, config)) {
    throw new Error(notSvelte(normalizedOptions.projectRoot));
  }

  const webPackageJsonPath = joinPathFragments(
    normalizedOptions.projectRoot,
    'package.json'
  );

  await initGenerator(tree, {
    skipFormat: normalizedOptions.skipFormat,
    tsConfigName: 'tsconfig.gb.json',
  });

  addNxConfig(tree, webPackageJsonPath);

  updatePrettier(tree, normalizedOptions);
  addWorkspaceToPackageJson(tree, normalizedOptions, 'package.json');

  normalizedOptions.eslint = false;
  if (normalizedOptions.eslint) {
    // TODO use @nx/eslint (@nx/eslint works with projects. this is not a project yet)
    updateEslint(tree, config);
    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@nx/eslint-plugin': NX_VERSION,
        '@typescript-eslint/eslint-plugin': typescriptESLintVersion,
        '@typescript-eslint/parser': typescriptESLintVersion,
        eslint: eslintVersion,
        'eslint-plugin-gb': eslintPluginGbVersion,
        'eslint-plugin-svelte': eslintPluginSvelteVersion,
      },
      'package.json'
    );
    tree.write(
      joinPathFragments(normalizedOptions.projectRoot, 'tsconfig.base.json'),
      '{ "extends": "./tsconfig.json" }'
    );
    addScriptsToPackageJson(tree, { lint: 'eslint .' }, webPackageJsonPath);
  }

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return async () => {
    installPackagesTask(tree);
  };
}
