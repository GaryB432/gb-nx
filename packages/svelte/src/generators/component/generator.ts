import {
  formatFiles,
  generateFiles,
  getProjects,
  names,
  readNxJson,
  updateNxJson,
  type Tree,
} from '@nx/devkit';
import { join } from 'path';
import {
  getSvelteConfig,
  getSvelteFiles,
  supportsRunes,
} from '../../utils/svelte';
import type { Schema as ComponentGeneratorSchema } from './schema';

export interface NormalizedComponentSchema extends ComponentGeneratorSchema {
  projectLibRoot: string;
  projectName: string;
}

function normalizeOptions(
  tree: Tree,
  options: ComponentGeneratorSchema
): NormalizedComponentSchema {
  const ws = readNxJson(tree);
  const projectName = options.project ?? ws?.defaultProject;
  const projects = getProjects(tree);
  if (!projectName) {
    throw new Error('no project');
  }
  const project = projects.get(projectName);

  if (!project) {
    throw new Error(`Project "${projectName}" not found`);
  }

  const runesSupport = supportsRunes(tree, project);

  if (options.runes && !runesSupport.supports) {
    throw new Error(
      `runes feature requires svelte >= 5 (currently '${runesSupport.svelte}')`
    );
  }

  const runes = options.runes ?? runesSupport.supports;

  const config = getSvelteConfig(tree, project);

  if (!config) {
    throw new Error(`Project "${projectName}" not configured for svelte`);
  }

  // TODO choose lib or routes (new sverdle has components next to route files)
  const { lib } = getSvelteFiles(config);
  options.language = options.language ?? 'js';
  options.style = options.style ?? 'css';
  return {
    ...options,
    runes,
    projectName,
    projectLibRoot: join(project.root, lib),
  };
}

function addFiles(tree: Tree, options: NormalizedComponentSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    tmpl: '',
  };
  generateFiles(
    tree,
    join(__dirname, './files'),
    join(options.projectLibRoot, options.directory ?? ''),
    templateOptions
  );
}

function addGeneratorDefaults(
  tree: Tree,
  options: NormalizedComponentSchema
): void {
  const nxJson = readNxJson(tree);

  if (!nxJson) {
    return;
  }

  const { directory, language, style } = options;
  nxJson.generators = nxJson.generators ?? {};
  nxJson.generators['@gb-nx/svelte:component'] = {
    directory,
    language,
    style,
    ...(nxJson.generators['@gb-nx/svelte:component'] || {}),
  };

  updateNxJson(tree, nxJson);
}

export default async function componentGenerator(
  tree: Tree,
  options: ComponentGeneratorSchema
): Promise<void> {
  const normalizedOptions = normalizeOptions(tree, options);
  addFiles(tree, normalizedOptions);
  addGeneratorDefaults(tree, normalizedOptions);
  if (!normalizedOptions.skipFormat) {
    await formatFiles(tree);
  }
}
