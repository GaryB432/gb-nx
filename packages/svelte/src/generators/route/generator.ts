import {
  formatFiles,
  getProjects,
  joinPathFragments,
  names,
  readNxJson,
  type ProjectConfiguration,
  type Tree,
} from '@nx/devkit';
import { getSvelteConfig, getSvelteFiles } from '../../utils/svelte';
import type { Schema } from './schema';

interface NormalizedSchema extends Schema {
  routePath: string;
}

function getRouteParam(typedParam: string): [string, string] {
  const r = typedParam.split('=', 2);
  return [r[0], r[1]];
}

function normalizeOptions(
  host: Tree,
  options: Schema
): Required<NormalizedSchema> {
  const ws = readNxJson(host);
  options.project = options.project ?? ws?.defaultProject;
  options.directory = options.directory ?? '';

  if (!options.project) {
    throw new Error('Project or defaultProject required');
  }

  const defaultOptions: Required<NormalizedSchema> = {
    directory: options.directory,
    language: 'js',
    load: 'none',
    name: options.name,
    project: options.project,
    skipFormat: false,
    skipTests: false,
    style: 'css',
    routePath: options.name,
  };
  return {
    ...defaultOptions,

    ...options,
    routePath: joinPathFragments(options.directory ?? '', options.name),
  };
}

interface Segment {
  isParam: boolean;
  paramType: string | undefined;
  path: string;
}

export function getSegments(
  options: Pick<NormalizedSchema, 'routePath'>
): Segment[] {
  return options.routePath.split('/').map<Segment>((part) => {
    const isParam = part.startsWith('[');
    const [path, paramType] = getRouteParam(
      isParam ? part.slice(1, part.length - 1) : part
    );

    return { isParam, path, paramType };
  });
}
function addLoadPage(
  tree: Tree,
  proj: ProjectConfiguration,
  locations: { routes: string },
  options: NormalizedSchema
): void {
  const segments = getSegments(options);
  const routeParams = segments.filter((s) => s.isParam).map((p) => p.path);
  const paramDeclaration =
    routeParams.length > 0
      ? `const { ${routeParams.join(', ')} } = params;`
      : '';

  const answer = segments
    .map((s) => (s.isParam ? `\${${s.path}}` : s.path))
    .join('/');

  switch (options.load) {
    case 'server': {
      const fname = joinPathFragments(
        proj.root,
        locations.routes,
        options.routePath,
        options.language === 'ts' ? '+page.server.ts' : '+page.server.js'
      );

      const ts = `import type { PageServerLoad, RouteParams } from './$types';
      async function demo(params: RouteParams): Promise<string> {
        ${paramDeclaration}
        return Promise.resolve(\`${answer}\`);
      }
      
      export const load = (async({ params }) => {
        return {
          subject: await demo(params)
        };
      }) satisfies PageServerLoad;`;

      const js = `async function demo(params) {
        ${paramDeclaration}
        return Promise.resolve('${options.name}');
      }
      /** @type {import('./$types').PageServerLoad} */
      export async function load({ params }) {
        return {
          subject: await demo(params)
        };
      }
      `;
      tree.write(fname, options.language === 'ts' ? ts : js);
      break;
    }
    case 'shared': {
      const fname = joinPathFragments(
        proj.root,
        locations.routes,
        options.routePath,
        options.language === 'ts' ? '+page.ts' : '+page.js'
      );

      const ts = `import type { PageLoad, RouteParams } from './$types';
      export const load = (async ({ params }) => {
        ${paramDeclaration}
        return {
          subject: \`${answer}\`
        };
      }) satisfies PageLoad;`;
      const js = `/** @type {import('./$types').PageLoad} */
      export function load({ params }) {
        ${paramDeclaration}
        return {
          subject: \`${answer}\`
        };
      }`;

      tree.write(fname, options.language === 'ts' ? ts : js);
      break;
    }
  }
}

function addSveltePage(
  tree: Tree,
  proj: ProjectConfiguration,
  locations: { routes: string },
  options: Required<NormalizedSchema>
): void {
  const fname = joinPathFragments(
    proj.root,
    locations.routes,
    options.routePath,
    '+page.svelte'
  );

  const scripts: Record<
    string,
    { none: string; server: string; shared: string }
  > = {
    js: {
      none: `<script>
        let data = { subject: '${options.name}' };
      </script>`,
      server: `<script>
        /** @type {import('./$types').PageData} */
        export let data;
      </script>`,
      shared: `<script>
        /** @type {import('./$types').PageData} */
        export let data;
      </script>`,
    },
    ts: {
      none: `<script lang="ts">
        let data = { subject: '${options.name}' };
      </script>`,
      server: `<script lang="ts">
        import type { PageData } from './$types';
        export let data: PageData;
      </script>`,
      shared: `<script lang="ts">
        import type { PageData } from './$types';
        export let data: PageData;
      </script>`,
    },
  };

  const script = scripts[options.language][options.load];

  const html = `<article class="container">
	{data.subject} works
</article>
`;

  // TODO add sass devdep in project if selected
  const styles = `<style${options.style === 'scss' ? ' lang="scss"' : ''}>
  .container {
    padding: 1em;
    border: thin solid silver;
  }
  @media screen and (min-width: 576px) { /* landscape phones */ }
  @media screen and (min-width: 768px) { /* tablets */ }
  @media screen and (min-width: 992px) { /* desktops */ }
  @media screen and (min-width: 1200px) { /* large desktops */ }
  @media screen and (min-width: 1400px) { /* larger desktops */ }
</style>`;

  tree.write(fname, [script, html, styles].join('\n\n'));
}
function addPlaywrightPage(
  tree: Tree,
  proj: ProjectConfiguration,
  locations: { tests: string },
  options: NormalizedSchema
): void {
  const fname = joinPathFragments(
    proj.root,
    locations.tests,
    options.directory ?? '',
    `${names(options.name).fileName}.spec.ts`
  );
  // const routePath = joinPathFragments(options.directory ?? '', options.name);

  const routePath = getSegments(options)
    .map((f) => (f.isParam ? `_${f.path}_` : f.path))
    .join('/');

  const content = `import { expect, test } from '@playwright/test';

test('has generated article', async ({ page }) => {
  await page.goto('/${routePath}');
  expect(
    await page.textContent('article.container')
  ).toContain('works');
});
`;
  tree.write(fname, content);
}

export async function routeGenerator(
  tree: Tree,
  schema: Schema
): Promise<void> {
  const notfound = (p: string) => `project '${p}' was not found in workspace`;
  const notSvelte = (p: string) =>
    `project '${p}' is not configured for svelte`;

  const options = normalizeOptions(tree, schema);

  const projects = getProjects(tree);
  const proj = projects.get(options.project);
  if (!proj) {
    throw new Error(notfound(options.project));
  }

  const configContent = getSvelteConfig(tree, proj);

  if (!configContent) {
    throw new Error(notSvelte(options.project));
  }

  const locations = getSvelteFiles(configContent);

  addLoadPage(tree, proj, locations, options);

  addSveltePage(tree, proj, locations, options);

  if (!options.skipTests) {
    addPlaywrightPage(tree, proj, { tests: 'tests' }, options);
  }
  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default routeGenerator;
