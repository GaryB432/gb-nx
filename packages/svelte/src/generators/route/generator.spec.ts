import { addProjectConfiguration, readNxJson, type Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { createSvelteKitApp } from '../../utils/svelte';
import generator, { getSegments } from './generator';
import { type Schema } from './schema';

import '../../utils/testing/matchers';

describe('route generator', () => {
  let appTree: Tree;

  beforeEach(() => {
    appTree = setupTreeWithSvelteProject('apps', 'test', '4.0.0');
  });
  describe('basics', () => {
    it('works with default values', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        project: 'test',
      });
      expect(
        appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
      ).toMatchSnapshot();

      expect(
        appTree.read('apps/test/tests/some-route.spec.ts', 'utf-8')
      ).toMatchSnapshot(); // TODO should be js
    });
    it('skips tests', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route-a',
        skipTests: true,
        project: 'test',
      });
      expect(appTree.children('apps/test/src/routes/some-route-a')).toEqual([
        '+page.svelte',
      ]);
      expect(appTree.children('apps/test/tests')).toEqual([]);
    });

    it('should add defaults to nx.json', async () => {
      await skipFormatGenerator(appTree, {
        name: 'hello',
        project: 'test',
        directory: 'a/b/c/d',
        language: 'ts',
        style: 'scss',
      });

      expect(readNxJson(appTree)!.generators!['@gb-nx/svelte:route']).toEqual({
        directory: 'a/b/c/d',
        language: 'ts',
        style: 'scss',
      });
    });
  });

  describe('--directory', () => {
    /* snapshots? */
    it('works with --directory="a/b/c"', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        directory: 'a/b/c',
        project: 'test',
      });
      expect(appTree.children('apps/test/src/routes/a/b/c/some-route')).toEqual(
        ['+page.svelte']
      );
      expect(appTree.children('apps/test/tests/a/b/c')).toEqual([
        'some-route.spec.ts', // TODO this should be js!
      ]);
    });
  });

  describe('--language', () => {
    /* snapshots? */
    it('works with --language=ts', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        language: 'ts',
        project: 'test',
      });
      expect(
        appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
      ).toMatchSnapshot();

      expect(
        appTree.read('apps/test/tests/some-route.spec.ts', 'utf-8')
      ).toMatchSnapshot();
    });
  });

  describe('--style', () => {
    /* snapshots? */

    it('works with --style=scss', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        language: 'ts',
        style: 'scss',
        project: 'test',
      });
      expect(appTree.children('apps/test/src/routes/some-route')).toEqual([
        '+page.svelte',
      ]);
      expect(
        appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
      ).toContain('<style lang="scss">');

      expect(appTree.children('apps/test/tests')).toEqual([
        'some-route.spec.ts',
      ]);
    });
  });

  describe('--load', () => {
    /* snapshots? */
    it('works with --load=shared', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        language: 'ts',
        load: 'universal',
        project: 'test',
      });
      expect(appTree.children('apps/test/src/routes/some-route')).toEqual([
        '+page.ts',
        '+page.svelte',
      ]);
      expect(
        appTree.read('apps/test/src/routes/some-route/+page.ts', 'utf-8')
      ).toMatchSnapshot();

      const pageSvelte = appTree.read(
        'apps/test/src/routes/some-route/+page.svelte',
        'utf-8'
      );
      expect(pageSvelte).not.toUseRunes();

      expect(appTree.children('apps/test/tests')).toEqual([
        'some-route.spec.ts',
      ]);
    });
    it('works with --load=server', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        language: 'ts',
        load: 'server',
        project: 'test',
      });
      expect(appTree.children('apps/test/src/routes/some-route')).toEqual([
        '+page.server.ts',
        '+page.svelte',
      ]);
      expect(
        appTree.read('apps/test/src/routes/some-route/+page.server.ts', 'utf-8')
      ).toMatchSnapshot();

      const pageSvelte = appTree.read(
        'apps/test/src/routes/some-route/+page.svelte',
        'utf-8'
      );
      expect(pageSvelte).not.toUseRunes();
      expect(pageSvelte).toContain('export let');

      expect(appTree.children('apps/test/tests')).toEqual([
        'some-route.spec.ts',
      ]);
    });
    it('works with --load=none', async () => {
      await skipFormatGenerator(appTree, {
        name: 'some-route',
        language: 'ts',
        load: 'none',
        project: 'test',
      });
      expect(appTree.children('apps/test/src/routes/some-route')).toEqual([
        '+page.svelte',
      ]);

      const pageSvelte = appTree.read(
        'apps/test/src/routes/some-route/+page.svelte',
        'utf-8'
      );
      expect(pageSvelte).not.toUseRunes();
      expect(pageSvelte).toContain('export let data');

      expect(appTree.children('apps/test/tests')).toEqual([
        'some-route.spec.ts',
      ]);
    });
  });

  describe('--runes', () => {
    describe('svelte 4', () => {
      beforeEach(() => {
        appTree = setupTreeWithSvelteProject('apps', 'test', '4.0.0');
      });
      it('works with --runes=undefined', async () => {
        await skipFormatGenerator(appTree, {
          name: 'some-route',
          project: 'test',
        });
        expect(
          appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
        ).toContain('let data = ');

        expect(appTree.children('apps/test/tests')).toEqual([
          'some-route.spec.ts',
        ]);
      });
      it('works with --no-runes', async () => {
        await skipFormatGenerator(appTree, {
          name: 'some-route',
          runes: false,
          project: 'test',
        });
        expect(
          appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
        ).toMatchSnapshot();

        expect(appTree.children('apps/test/tests')).toEqual([
          'some-route.spec.ts',
        ]);
      });
      it('works with --runes', async () => {
        //

        expect(async () => {
          await skipFormatGenerator(appTree, {
            name: 'some-route',
            runes: true,
            project: 'test',
          });
        }).rejects.toThrow(
          "runes feature requires svelte >= 5 (currently '4.0.0')"
        );
      });
    });

    describe('svelte 5', () => {
      beforeEach(() => {
        appTree = setupTreeWithSvelteProject('apps', 'test', '5.0.0');
      });
      it('works with --runes=undefined', async () => {
        await skipFormatGenerator(appTree, {
          name: 'some-route',
          project: 'test',
        });
        expect(
          appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
        ).toContain('let data = ');

        expect(appTree.children('apps/test/tests')).toEqual([
          'some-route.spec.ts',
        ]);
      });
      it('works with --no-runes', async () => {
        await skipFormatGenerator(appTree, {
          name: 'some-route',
          runes: false,
          project: 'test',
        });
        expect(
          appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
        ).toMatchSnapshot();

        expect(appTree.children('apps/test/tests')).toEqual([
          'some-route.spec.ts',
        ]);
      });
      it('works with --runes', async () => {
        await skipFormatGenerator(appTree, {
          name: 'some-route',
          runes: true,
          project: 'test',
        });
        expect(
          appTree.read('apps/test/src/routes/some-route/+page.svelte', 'utf-8')
        ).toUseRunes();
      });
    });
  });

  describe('getSegments', () => {
    it('gets segments', () => {
      expect(getSegments({ routePath: '[just]' })).toEqual([
        { path: 'just', paramType: undefined, isParam: true },
      ]);
      expect(getSegments({ routePath: '[ab]/cd/ef/[gh=atype]/ij' })).toEqual([
        { path: 'ab', paramType: undefined, isParam: true },
        { path: 'cd', paramType: undefined, isParam: false },
        { path: 'ef', paramType: undefined, isParam: false },
        { path: 'gh', paramType: 'atype', isParam: true },
        { path: 'ij', paramType: undefined, isParam: false },
      ]);
    });
  });
});

function skipFormatGenerator(tree: Tree, options: Schema): Promise<void> {
  return generator(tree, { ...options, skipFormat: true });
}

function setupTreeWithSvelteProject(
  directory: string,
  name: string,
  svelteVersion: string
): Tree {
  const appTree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  createSvelteKitApp(appTree, svelteVersion, {
    directory,
    name,
    skipFormat: true,
  });
  const root = directory.concat('/').concat(name);
  addProjectConfiguration(appTree, name, {
    root,
  });
  return appTree;
}
