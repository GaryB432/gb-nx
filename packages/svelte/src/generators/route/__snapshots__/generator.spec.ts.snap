// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route generator --language works with --language=ts 1`] = `
"<script lang="ts">
        export let data = { subject: 'some-route' };
        </script>

<svelte:head>
  <title>test - some-route</title>
</svelte:head>

<article class="container">
	{data.subject} works
</article>


<style>
  .container {
    padding: 1em;
    border: thin solid silver;
  }
  @media screen and (min-width: 576px) { /* landscape phones */ }
  @media screen and (min-width: 768px) { /* tablets */ }
  @media screen and (min-width: 992px) { /* desktops */ }
  @media screen and (min-width: 1200px) { /* large desktops */ }
  @media screen and (min-width: 1400px) { /* larger desktops */ }
</style>"
`;

exports[`route generator --language works with --language=ts 2`] = `
"import { expect, test } from '@playwright/test';

test('has generated article', async ({ page }) => {
  await page.goto('/some-route');
  expect(
    await page.textContent('article.container')
  ).toContain('works');
});
"
`;

exports[`route generator --load works with --load=server 1`] = `
"import type { PageServerLoad, RouteParams } from './$types';
      async function demo(params: RouteParams): Promise<string> {
        
        return Promise.resolve(\`some-route\`);
      }
      
      export const load = (async ({ params }) => {
        return {
          subject: await demo(params)
        };
      }) satisfies PageServerLoad;
      "
`;

exports[`route generator --load works with --load=shared 1`] = `
"import type { PageLoad, RouteParams } from './$types';
      export const load = (async ({ params }) => {
        
        return {
          subject: \`some-route\`
        };
      }) satisfies PageLoad;"
`;

exports[`route generator --runes svelte 4 works with --no-runes 1`] = `
"<script>
        let data = { subject: 'some-route' };
      </script>

<svelte:head>
  <title>test - some-route</title>
</svelte:head>

<article class="container">
	{data.subject} works
</article>


<style>
  .container {
    padding: 1em;
    border: thin solid silver;
  }
  @media screen and (min-width: 576px) { /* landscape phones */ }
  @media screen and (min-width: 768px) { /* tablets */ }
  @media screen and (min-width: 992px) { /* desktops */ }
  @media screen and (min-width: 1200px) { /* large desktops */ }
  @media screen and (min-width: 1400px) { /* larger desktops */ }
</style>"
`;

exports[`route generator --runes svelte 5 works with --no-runes 1`] = `
"<script>
        let data = { subject: 'some-route' };
      </script>

<svelte:head>
  <title>test - some-route</title>
</svelte:head>

<article class="container">
	{data.subject} works
</article>


<style>
  .container {
    padding: 1em;
    border: thin solid silver;
  }
  @media screen and (min-width: 576px) { /* landscape phones */ }
  @media screen and (min-width: 768px) { /* tablets */ }
  @media screen and (min-width: 992px) { /* desktops */ }
  @media screen and (min-width: 1200px) { /* large desktops */ }
  @media screen and (min-width: 1400px) { /* larger desktops */ }
</style>"
`;

exports[`route generator basics works with default values 1`] = `
"<script>
        let data = { subject: 'some-route' };
      </script>

<svelte:head>
  <title>test - some-route</title>
</svelte:head>

<article class="container">
	{data.subject} works
</article>


<style>
  .container {
    padding: 1em;
    border: thin solid silver;
  }
  @media screen and (min-width: 576px) { /* landscape phones */ }
  @media screen and (min-width: 768px) { /* tablets */ }
  @media screen and (min-width: 992px) { /* desktops */ }
  @media screen and (min-width: 1200px) { /* large desktops */ }
  @media screen and (min-width: 1400px) { /* larger desktops */ }
</style>"
`;

exports[`route generator basics works with default values 2`] = `
"import { expect, test } from '@playwright/test';

test('has generated article', async ({ page }) => {
  await page.goto('/some-route');
  expect(
    await page.textContent('article.container')
  ).toContain('works');
});
"
`;
